src/poke_env/environment/abstract_battle.py

```diff
@@ -210,6 +210,54 @@ def get_pokemon(
     def _clear_all_boosts(self):  # pragma: no cover
         pass
 
+    def _check_damage_message_for_item(self, split_message: List[str]):
+        # Catches when a side takes damage from the opponent's item
+        # The item belongs to the side not taking damage
+        # Example:
+        #   |-damage|p2a: Archeops|88/100|[from] item: Rocky Helmet|[of] p1a: Ferrothorn
+        if (
+            len(split_message) == 6
+            and split_message[4].startswith("[from] item:")
+            and split_message[5].startswith("[of]")
+        ):
+            item = split_message[4].split("item:")[-1]
+            pkmn = split_message[5].split("[of]")[-1].strip()
+            self.get_pokemon(pkmn)._item = to_id_str(item)
+
+        # Catches when a side takes damage from it's own item
+        # The item belongs to the same side taking damage
+        # Example:
+        #   |-damage|p2a: Pikachu|90/100|[from] item: Life Orb
+        elif len(split_message) == 5 and split_message[4].startswith("[from] item:"):
+            item = split_message[4].split("item:")[-1]
+            pkmn = split_message[2]
+            self.get_pokemon(pkmn)._item = to_id_str(item)
+
+    def _check_damage_message_for_ability(self, split_message: List[str]):
+        # Catches when a side takes damage from the opponent's ability
+        # the item is from the side not taking damage
+        # Example:
+        #   |-damage|p2a: Archeops|88/100|[from] ability: Iron Barbs|[of] p1a: Ferrothorn
+        if (
+            len(split_message) == 6
+            and split_message[4].startswith("[from] ability:")
+            and split_message[5].startswith("[of]")
+        ):
+            ability = split_message[4].split("ability:")[-1]
+            pkmn = split_message[5].split("[of]")[-1].strip()
+            self.get_pokemon(pkmn)._ability = to_id_str(ability)
```

Comment:
`ability` also has a setter - I'd change it to `self.get_pokemon(pkmn).ability = to_id_str(ability)`